libs/MobileSync/src/com/salesforce/androidsdk/mobilesync/manager/LayoutSyncManager.kt

```diff
@@ -0,0 +1,335 @@
+            if ((results == null) || (results.length() == 0)) {
+                if (fallbackOnServer) {
+                    fetchFromServer(
+                        objectAPIName,
+                        formFactor,
+                        layoutType,
+                        mode,
+                        recordTypeId,
+                        syncCallback
+                    )
+                } else {
+                    onSyncComplete(
+                        objectAPIName,
+                        formFactor,
+                        layoutType,
+                        mode,
+                        recordTypeId,
+                        syncCallback,
+                        null
+                    )
+                }
+            } else {
+                onSyncComplete(
+                    objectAPIName, formFactor, layoutType, mode, recordTypeId, syncCallback,
+                    Layout.fromJSON(results.optJSONArray(0).optJSONObject(0))
+                )
+            }
+        } catch (e: Exception) {
+            MobileSyncLogger.e(
+                TAG,
+                "Exception occurred while reading layout data from the cache",
+                e
+            )
+        }
+    }
+
+    private fun onSyncComplete(
+        objectAPIName: String, formFactor: String, layoutType: String, mode: String,
+        recordTypeId: String?, syncCallback: LayoutSyncCallback?, layout: Layout?
+    ) {
+        syncCallback?.onSyncComplete(
+            objectAPIName,
+            formFactor,
+            layoutType,
+            mode,
+            recordTypeId,
+            layout
+        )
+    }
+
+    private fun initializeSoup() {
+        if (!smartStore.hasSoup(SOUP_NAME)) {
+            smartStore.registerSoup(SOUP_NAME, INDEX_SPECS)
+        }
+    }
+
+    /**
+     * Callback interface for layout sync.
+     *
+     * @author bhariharan
+     */
+    interface LayoutSyncCallback {
+        /**
+         * Callback triggered when layout sync completes.
+         *
+         * @param objectAPIName Object API name.
+         * @param formFactor Form factor.
+         * @param layoutType Layout type.
+         * @param mode Mode.
+         * @param recordTypeId Record type ID.
+         * @param layout Layout.
+         */
+        fun onSyncComplete(
+            objectAPIName: String?, formFactor: String?, layoutType: String?,
+            mode: String?, recordTypeId: String?, layout: Layout?
+        )
+    }
+
+    companion object {
+        const val SOUP_NAME = "sfdcLayouts"
+        const val QUERY = "SELECT {" + SOUP_NAME + ":_soup} FROM {" + SOUP_NAME +
+                "} WHERE {" + SOUP_NAME + ":" + Constants.ID + "} = '%s-%s-%s-%s-%s'"
+        private const val TAG = "LayoutSyncManager"
+        private val INDEX_SPECS = arrayOf(
+            IndexSpec(Constants.ID, SmartStore.Type.json1)
+        )
+        private val INSTANCES: MutableMap<String, LayoutSyncManager> = HashMap()
```

Comment:
No Change Required: I was curious to double check: Kotlin's `MutableMap` isn't thread safe like our old friend `HashMap`.  So this might not be a candidate for `mutableMapOf`.  So this probably needs to stay as-is.